### DISCUSSION TITLE
Need some help with bulk updates (async)

### DISCUSSION THREAD

**Discussion by vamshiaruru-virgodesigns** [sqlalchemy/sqlalchemy#8801] [Usage Questions] [answered]
Hi everyone, I am trying to do some bulk updates and I am not sure what I am doing wring. I am trying to update existing rows with new values (each row gets different values for updates).

Here's what I tried
```python
query = select(Event)
async with AsyncSessionLocal() as session:
    resp = await session.execute(query)
    resp = resp.scalars().all()
async with AsyncSessionLocal() as session:
    resp = await session.execute(update(Event), [{"id": resp[0].id, "event_type": resp[0].event_type}])
```
In a real bulk update scenario, the array would have a lot of elements.  I get the following error:
```
IntegrityError: (sqlalchemy.dialects.postgresql.asyncpg.IntegrityError) <class 'asyncpg.exceptions.UniqueViolationError'>: duplicate key value violates unique constraint "event_event_pkey"
DETAIL:  Key (id)=(99b50163-bdb0-444d-bab2-22cde20ba70d) already exists.
[SQL: UPDATE event_event SET id=%s, event_type=%s]
[parameters: (UUID('99b50163-bdb0-444d-bab2-22cde20ba70d'), 'tag_updated')]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
```
In this example I was able to get the entire row from DB but in the actual scenario, I just have ids and updates corresponding to those ids, and I'd prefer not to fetch all those rows before updating.
What am I doing wrong? Any help is appreciated, thank you :)

**Comment by zzzeek** [comment, marked as answer]
hi -

it looks like you are using the 2.0 feature, but that's not implemented in 1.4.    For 1.4 you would stick with [bulk_update_mappings](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.bulk_update_mappings).

if OTOH that's 2.0 then we need more information.

**Reply by vamshiaruru-virgodesigns** [reply]
Thanks for the response, I am using `sqlalchemy = "^1.4.39"`. I tried using the `bulk_update_mapping` method, but I am getting the following error:
```
AttributeError: 'AsyncSession' object has no attribute 'bulk_update_mappings'
```
Here's how I tried to test it 
```python
engine = create_async_engine(
    settings.database_dsn,
    poolclass=NullPool,
    pool_pre_ping=True,
)

AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)  # type: ignore
async with AsyncSessionLocal() as session:
    print(session.bulk_update_mappings)
```
Looks like the method is not exposed on async sessions?

**Reply by zzzeek** [reply]
oh.    right yeah :)   it's not on async, because it's legacy.

if you just want to run an UPDATE statement, you can do it mostly as you are just set up the WHERE criteria like this, probably run it through connection():

```py
conn = await session.connection()
resp = await conn.execute(update(Event).where(Event.id==bindparam("event_id")), 
     [{"event_id": resp[0].id, "event_type": resp[0].event_type}])
```

**Reply by vamshiaruru-virgodesigns** [reply]
Thanks! I'll try that out in some time and get back to you :)

**Reply by vamshiaruru-virgodesigns** [reply]
Hi, just tried it out. It works, thanks!

**Reply by vamshiaruru-virgodesigns** [reply]
Out of curiosity, is is possible to see that queries sqlalchemy is omitting for this bulk update?

**Reply by CaselIT** [reply]
like all other engine logging the easiest way is to add `echo=True` to the `create_engine` call

**Reply by hilmanski** [reply]
hi @zzzeek is that async bulk update method (with bindparam) is the best we have for <= 1.4 version? I mean to ask in terms of performance. Thanks.

**Reply by zzzeek** [reply]
it runs just one UPDATE statement on the DB, so if that's the operation you need to do, that will perform the best.    synchronize_session=False will eliminate any attempts to reconcile local objects with the UPDATE which can in some scenarios add to the overall time (usually not much).

**Reply by hilmanski** [reply]
thanks a lot @zzzeek
